### PR TITLE
[Merged by Bors] - feat(GroupTheory/GroupAction): define MulAction.period and create GroupTheory/GroupAction/Period

### PR DESCRIPTION
**Pull request by adri326**
Defines `MulAction.period g a` as a wrapper around `Function.minimalPeriod (fun x => g • x) a`,
allowing for some cleaner proofs involving the period of a group action.
The existing `MulAction.*_minimalPeriod_*` lemmas are changed to be defined using their now-preferred `MulAction.*_period_*` counterparts,
although they will only be made deprecated in another pull request.
The `Mathlib.GroupTheory.GroupAction.Period` module is also created, for additional lemmas around `MulAction.period`.

Some core lemmas need to remain in `Mathlib.Dynamics.PeriodicPts`, as they are needed for `ZMod` and the quotient group.

---

Note that the linter fails on `Mathlib.Dynamic.PeriodicPts`, since I do not know how to split the `to_additive` description on multiples lines :/

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
> Note that the linter fails on Mathlib.Dynamic.PeriodicPts, since I do not know how to split the to_additive description on multiples lines :/
I think a newline without indentation works, i.e.
```
@[to_additive "foo
bar"]
def 
```

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (erdOne) [inline Mathlib/Dynamics/PeriodicPts.lean:681]:
```suggestion
  rw [period_eq_minimalPeriod, ← isPeriodicPt_iff_minimalPeriod_dvd, fixed_iff_isPeriodicPt]
```

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (erdOne) [inline Mathlib/Dynamics/PeriodicPts.lean:693]:
```suggestion
  · rw [Int.negSucc_coe, zpow_neg, zpow_ofNat, inv_smul_eq_iff, eq_comm, dvd_neg, Int.coe_nat_dvd,
      pow_smul_eq_iff_period_dvd]
```

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (erdOne) [inline Mathlib/Dynamics/PeriodicPts.lean:631]: I think you can remove a lot of `Function.` thanks to this.

**Comment by adri326** [comment]
Oh it does, thanks

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (erdOne) [inline Mathlib/Dynamics/PeriodicPts.lean:710]:
```suggestion
  conv_rhs => rw [← Nat.mod_add_div n (period m a), pow_smul_plus_period]
```

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (erdOne) [inline Mathlib/Dynamics/PeriodicPts.lean:717]:
```suggestion
  conv_rhs => rw [← Int.emod_add_ediv j (period g a), zpow_smul_plus_period]
```

**Review by erdOne — COMMENTED** [review]
Some stylistic comments

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (adri326) [reply Mathlib/Dynamics/PeriodicPts.lean:631]: True, though I wonder if accidental name shadowing hasn't happened in the past

**Comment by adri326** [comment]
Fixed the style issues pointed out above :)

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (erdOne) [reply Mathlib/Dynamics/PeriodicPts.lean:631]: I think lean will complain about ambiguous names.

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (adri326) [reply Mathlib/Dynamics/PeriodicPts.lean:711]: This is what `pow_smul_eq_iff_period_dvd` does above, these theorems are left here for compatibility purposes, since they're used in defining quotient groups, and will be made deprecated in #9514.

Comment (tb65536) [reply Mathlib/Dynamics/PeriodicPts.lean:711]: Ah, great.

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (Vierkantor) [inline Mathlib/Dynamics/PeriodicPts.lean:705]: I'd like to restructure this section somewhat: IMO the fundamental relation is that `m ^ k • a = m ^ n • a` if `n ≡ k` mod `period m a`. But we can't express that here because `Nat.ModEq` and `Int.ModEq` are not available, so the next best thing is to put the `mod` lemmas first.

Then I'm going to derive the addition results from them, and include addition on the left and on the right. Finally I'm renaming them since `+` is called `add` here, not `plus`. :)
```suggestion
@[to_additive (attr := simp)]
theorem pow_smul_mod_period (n : ℕ) {m : M} {a : α}:
    m ^ (n % period m a) • a = m ^ n • a := by
  conv_rhs => rw [← Nat.mod_add_div n (period m a), pow_add, mul_smul,
    pow_smul_eq_iff_period_dvd.mpr (dvd_mul_right _ _)]

@[to_additive (attr := simp)]
theorem zpow_smul_mod_period (j : ℤ) {g : G} {a : α}:
    g ^ (j % (period g a : ℤ)) • a = g ^ j • a := by
  conv_rhs => rw [← Int.emod_add_ediv j (period g a), zpow_add, mul_smul,
    zpow_smul_eq_iff_period_dvd.mpr (dvd_mul_right _ _)]

@[to_additive (attr := simp)]
theorem pow_add_period_smul (n : ℕ) (m : M) (a : α):
    m ^ (n + period m a) • a = m ^ n • a := by
  rw [← pow_smul_mod_period, Nat.add_mod_right, pow_smul_mod_period]

@[to_additive (attr := simp)]
theorem pow_period_add_smul (n : ℕ) (m : M) (a : α):
    m ^ (period m a + n) • a = m ^ n • a := by
  rw [← pow_smul_mod_period, Nat.add_mod_left, pow_smul_mod_period]

@[to_additive (attr := simp)]
theorem zpow_add_period_smul (i : ℤ) (g : G) (a : α):
    g ^ (i + period g a) • a = g ^ i • a := by
  rw [← zpow_smul_mod_period, Int.add_emod_self, zpow_smul_mod_period]

@[to_additive (attr := simp)]
theorem zpow_period_add_smul (i : ℤ) (g : G) (a : α):
    g ^ (period g a + i) • a = g ^ i • a := by
  rw [← zpow_smul_mod_period, Int.add_emod_self_left, zpow_smul_mod_period]
```

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (Vierkantor) [inline Mathlib/Dynamics/PeriodicPts.lean:712]: Nice cleanup!

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:9]: My hobby is sorting imports alphabetically :)
```suggestion
import Mathlib.Dynamics.PeriodicPts
import Mathlib.GroupTheory.Exponent
import Mathlib.GroupTheory.GroupAction.Basic
```

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:31]: Theorem names really have to closely match their statement, so:
```suggestion
theorem le_period_of_moved {m : M} {a : α} {n : ℕ} (period_pos : 0 < period m a)
```

(I assume "moved" is used as the opposite of "fixed (point)"? I'm not familiar with any word for this concept, so maybe another name would be clearer for other users too.)

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:36]: `by_contra!` can do some of the rewrites for you automatically:
```suggestion
  by_contra! period_lt_n
  apply moved _ period_pos period_lt_n
  exact smul_pow_period_fixed m a
```

Comment (Vierkantor) [reply Mathlib/GroupTheory/GroupAction/Period.lean:36]: In fact if you really love term mode this could even become:
```suggestion
  le_of_not_gt <| fun period_lt_n => moved _ period_pos period_lt_n <| smul_pow_period_fixed m a
```
(although I would not advise it!)

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:44]: Here I would not go into tactic mode:
```suggestion
    period m a ≤ n :=
  (isPeriodicPt_smul_iff.mpr fixed).minimalPeriod_le n_pos
```

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:52]: Idem:
```suggestion
    0 < period m a :=
  (isPeriodicPt_smul_iff.mpr fixed).minimalPeriod_pos n_pos
```

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:60]: There's a little trick here: for `n : ℕ`, `0 < n` is definitionally equal to `1 ≤ n`.
```suggestion
theorem period_eq_one_of_fixed {m : M} {a : α} (fixed : m • a = a) : period m a = 1 := :=
  le_antisymm
    (period_le_of_fixed one_pos (by simpa))
    (period_pos_of_fixed one_pos (by simpa))
```

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:68]: All these tactic steps can also be term steps. I'd also like to replace the specific `Nat.not_le` with a more general `not_le_of_gt`.
```suggestion
    m^n • a ≠ a :=
  fun a_fixed => not_le_of_gt n_lt_period (period_le_of_fixed n_pos a_fixed)
```

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:108]: Following a plain `simp` with another tactic ("non-terminal `simp`") is dangerous since the set of `simp` lemmas can change and break the proof. `simp` followed by `apply`/`exact` can often be combined into `simpa` instead:
```suggestion
  simpa [upperBounds] using period_le_exponent exp_pos _
```

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (Vierkantor) [inline Mathlib/Dynamics/PeriodicPts.lean:645]: I think we should add a few more lemmas about special cases of `period`, in particular `period_one : period 1 x = 1` and `period_inv : period g⁻¹ x = period g x`? (Assuming those are correct!)

**Review by Vierkantor — COMMENTED** [review]
Thanks for the PR. It might look like I have a lot of comments, but I'm confident that they'll be the (almost) complete set, once they're done everything is ready to merge. This PR is nicely structured overall!

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (adri326) [reply Mathlib/GroupTheory/GroupAction/Period.lean:9]: I should try to automate that :p

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (adri326) [reply Mathlib/Dynamics/PeriodicPts.lean]: It's unfortunate that `rw` doesn't provide an easy way to do eta-expansion for existing lemmas like `smul_iterate`

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (adri326) [reply Mathlib/GroupTheory/GroupAction/Period.lean:31]: To be fair, this is the only real mechanism to prove a lower bound for the period, so we could just go with `le_period`

**Comment by adri326** [comment]
Couple of changes, as suggested; of interest:
- `le_period_of_moved` → `le_period`
- `moved_of_lt_period` → `pow_smul_ne_of_lt_period` (currently lives in the `MulAction` namespace)
- `period_eq_one_of_fixed` → `period_eq_one_iff`
- new `period_one` and `period_inv` (I couldn't golf the latter any further)
- new `period_dvd_orderOf` and `period_dvd_exponent`, which helped golfed the rest of the lemmas
- added missing docstrings in `@[to_additive]`

File: Mathlib/Dynamics/PeriodicPts.lean
Comment (Vierkantor) [inline Mathlib/Dynamics/PeriodicPts.lean:647]: Good point about `to_additive`izing the doc comments, let's not forget this one!
```suggestion
/-- `MulAction.period m a` is definitionally equal to `Function.minimalPeriod (m • ·) a`. -/
@[to_additive "`AddAction.period m a` is definitionally equal to `Function.minimalPeriod (m +ᵥ ·) a`"]
```

File: Mathlib/GroupTheory/GroupAction/Period.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/GroupAction/Period.lean:53]: For some reason the style guide says brackets belong on the next line, even if it makes indentation look slightly ugly.
```suggestion
theorem period_eq_one_iff {m : M} {a : α} : period m a = 1 ↔ m • a = a :=
  ⟨fun eq_one => pow_one m ▸ eq_one ▸ pow_period_smul m a,
   fun fixed => le_antisymm
    (period_le_of_fixed one_pos (by simpa))
    (period_pos_of_fixed one_pos (by simpa))⟩
```

**Review by Vierkantor — APPROVED** [review]
Some minor comments and I'm happy with this, thanks! I'm going to ask another maintainer to delegate this because I haven't reviewed `Dynamics` before and I contributed a few of the lemmas.

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
